aten/src/ATen/native/Math.h

```diff
@@ -384,3 +384,120 @@ calc_gcd(T a, T b) {
   }
   return b;
 }
+
+/*
+ * The following function has been modified from code that comes with the following copyright notice.
```

Comment:
There are multiple distributions of Cephes and its author, Stephen Moshier, has made clear that the code can be reused freely (and without requiring a particular license or copyright be distributed with it). I would add a comment document this function as normal, then add a sentence stating it's derived from an implementation in the Cephes Math Library.

Reply:
I actually took some time to compare scipy.special.i0 with an approximation of the modified Bessel function of the first kind that's accurate in the limit, and I was impressed with how closely they matched on both smaller and larger values. In particular, on larger values I had to go out to 40 iterations of the approximating series before approach the value produced by scipy.special.i0! I also compared two more modern and simpler approximations that try to fit the modified Bessel function of the first kind and was disappointed in how poorly they approximated the function.

Reply:
So then I read this blog post: https://www.advanpix.com/2015/11/11/rational-approximations-for-the-modified-bessel-function-of-the-first-kind-i0-computations-double-precision/#id2020094592. 

This was actually implemented in a modified form in Boost (see code links in the comments). I implemented their approach in Python (the correct argument is `x * x / 4` for the small argument approximation, despite the blog's Latex showing a ceil) and was very impressed with its accuracy. 

I actually really like this method because its derivation is clear and it's adaptable to different float types. How interested are you in the numerics? I think it'd be really interesting to investigate this approach further.

Reply:
The copyright notice for Cephes we should use is:

> Copyright (c) 2018, Steven Moshier
> All rights reserved.
> 
> Redistribution and use in source and binary forms, with or without
> modification, are permitted provided that the following conditions are met:
>     * Redistributions of source code must retain the above copyright
>       notice, this list of conditions and the following disclaimer.
>     * Redistributions in binary form must reproduce the above copyright
>       notice, this list of conditions and the following disclaimer in the
>       documentation and/or other materials provided with the distribution.
>     * Neither the name of the <organization> nor the
>       names of its contributors may be used to endorse or promote products
>       derived from this software without specific prior written permission.
> 
> THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
> ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
> WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
> DISCLAIMED. IN NO EVENT SHALL Steven Moshier BE LIABLE FOR ANY
> DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
> (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
> LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
> ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
> (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
> SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

Let's put this license in a note:

```
// Note [3-Clause BSD License for the Cephes Math Library]
// Code derived from implementations in the Cephes Math Library should mention its derivation and reference 
// this note (ex. 'This function is derived from the implementation of X in the Cephes Math Library. See note
// [3-Clause BSD License for the Cephes Math Library]. The license is:
// <license goes here>
```

Then each function can just explain its derivation and reference the note.

Reply:
Where would I place this note (what directory/file)? Also, should I change the comments on the zeta and digamma functions to follow this reference style for cephes?

Reply:
You can replace this first comment in ATen/native/Math.h with it.

Thank you for asking. Yes, it'd be great if you would update the other Cephes references. Looks like there are existing references in ATen/native/cuda/Math.cuh, ATen/native/Math.h, and ATen/native/Distributions.h.